File: DataCollector/src/main/java/com/braintreepayments/api/datacollector/DataCollectorInternalRequest.java
package com.braintreepayments.api.datacollector;

import java.util.HashMap;

import androidx.annotation.NonNull;
import androidx.annotation.RestrictTo;

/**
 * Used to configuration the PayPalDataCollector request
 */
@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
public class DataCollectorInternalRequest {

    private String applicationGuid;
    private String clientMetadataId;
    private boolean disableBeacon;
    private boolean hasUserLocationConsent;
    private HashMap<String,String> additionalData;

    DataCollectorInternalRequest(boolean hasUserLocationConsent) {
        this.hasUserLocationConsent = hasUserLocationConsent;
    }

    /**
     * @param additionalData Additional data that should be associated with the data collection.
     */
    DataCollectorInternalRequest setAdditionalData(HashMap<String, String> additionalData) {
        this.additionalData = additionalData;

        return this;
    }

    /**
     * @param applicationGuid The application global unique identifier.
     *                        There is a 36 character length limit on this value.
     */
    DataCollectorInternalRequest setApplicationGuid(String applicationGuid) {
        this.applicationGuid = applicationGuid;

        return this;
    }

    /**
     * @param riskCorrelationId The desired pairing ID, trimmed to 32 characters.
     */
    DataCollectorInternalRequest setRiskCorrelationId(@NonNull String riskCorrelationId) {
        this.clientMetadataId = riskCorrelationId.substring(0, Math.min(riskCorrelationId.length(), 32));

        return this;
    }

    /**
     * @param disableBeacon Indicates if the beacon feature should be disabled.
     */
    DataCollectorInternalRequest setDisableBeacon(boolean disableBeacon) {
        this.disableBeacon = disableBeacon;

        return this;
    }

    HashMap<String, String> getAdditionalData() {
        return additionalData;
    }

    String getApplicationGuid() {
        return applicationGuid;
    }

    String getClientMetadataId() {
        return clientMetadataId;
    }

    public boolean getHasUserLocationConsent() {
        return hasUserLocationConsent;
    }

    boolean isDisableBeacon() {
        return disableBeacon;
    }
}

File: DataCollector/src/main/java/com/braintreepayments/api/datacollector/DataCollectorInternalRequest.kt
package com.braintreepayments.api.datacollector

import androidx.annotation.RestrictTo

/**
 * Used to configuration the PayPalDataCollector request
 *
 * @property hasUserLocationConsent Whether the request has user location consent
 * @Property additionalData Additional data that should be associated with the data collection.
 * @Property applicationGuid The application global unique identifier.
 * There is a 36 character length limit on this value.
 * @Property isDisableBeacon Indicates if the beacon feature should be disabled.
 */
@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
data class DataCollectorInternalRequest(
    val hasUserLocationConsent: Boolean,
    var additionalData: HashMap<String, String>? = null,
    var applicationGuid: String? = null,
    var isDisableBeacon: Boolean = false
) {

    /**
     * @Property clientMetadataId The desired pairing ID, trimmed to 32 characters.
     */
    var clientMetadataId: String? = null
        set(value) {
            field = value?.let {
                if (it.length > RISK_CORRELATION_ID_MAX_LENGTH) {
                    it.substring(0, RISK_CORRELATION_ID_MAX_LENGTH)
                } else {
                    it.substring(0, it.length)
                }
            }
        }

    companion object {
        private const val RISK_CORRELATION_ID_MAX_LENGTH = 32
    }
}
